components/feature/webnotifications/src/main/java/mozilla/components/feature/webnotifications/WebNotificationFeature.kt

```diff
@@ -0,0 +1,103 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.webnotifications
+
+import android.app.NotificationChannel
+import android.app.NotificationManager
+import android.content.Context
+import android.os.Build
+import androidx.core.app.NotificationCompat
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import mozilla.components.browser.icons.BrowserIcons
+import mozilla.components.concept.engine.Engine
+import mozilla.components.concept.engine.webnotifications.WebNotification
+import mozilla.components.concept.engine.webnotifications.WebNotificationDelegate
+import mozilla.components.support.base.log.logger.Logger
+import java.lang.UnsupportedOperationException
+
+private const val NOTIFICATION_CHANNEL_ID = "mozac.feature.webnotifications.generic"
+
+/**
+ * Feature implementation for configuring and displaying web notifications to the user.
+ *
+ * Initialize this feature globally once on app start
+ * ```Kotlin
+ * WebNotificationFeature(
+ *     applicationContext, engine, icons, R.mipmap.ic_launcher, BrowserActivity::class.java
+ * )
+ * ```
+ *
+ * @param context The application Context.
+ * @param engine The browser engine.
+ * @param browserIcons The entry point for loading the large icon for the notification.
+ * @param launcherIcon The small icon for the notification.
+ * @param activityClass The Activity that the notification will launch if user taps on it
+ */
+class WebNotificationFeature(
+    val context: Context,
+    val engine: Engine,
+    val browserIcons: BrowserIcons,
+    val launcherIcon: Int,
+    val activityClass: Class<*>?
+) {
+    private val logger = Logger("AutoPushFeature")
+    private var pendingRequestId = 0
+    private var notificationId = 0
+    private val notificationIdMap = HashMap<String, Int>()
+
+    init {
+        try {
+            engine.registerWebNotificationDelegate(FeatureWebNotificationDelegate())
+        } catch (e: UnsupportedOperationException) {
+            logger.error("failed to register for web notification delegate", e)
+        }
+    }
+
+    private inner class FeatureWebNotificationDelegate : WebNotificationDelegate {
```

Comment:
Since `WebNotificationFeature` doesn't have any methods, should it implement `WebNotificationDelegate` and contain all these functions instead of having an inner class?

Reply:
Agreed.  I'll just extend `WebNotificationDelegate` with `WebNotificationFeature`